cclient/http/http.c

```diff
@@ -0,0 +1,231 @@
+  char buffer[RECEIVE_BUFFER_SIZE] = {0};
+  ssize_t num_received = 0;
+  // Setup parser settings - callbacks
+  http_parser_settings settings = {0};
+  settings.on_headers_complete = &request_parse_header_complete_cb;
+  settings.on_body = &request_parse_data_cb;
+  settings.on_message_complete = &request_parse_message_complete_cb;
+  // Setup parser
+  http_parser parser = {0};
+  // Setup response structure
+  struct _response_ctx response_context = {0};
+  response_context.parser = &parser;
+  response_context.response = response;
+  response_context.status = IOTA_REQUEST_STATUS_OK;
+  response_context.offset = 0;
+  // Initialize parser
+  http_parser_init(&parser, HTTP_RESPONSE);
+  parser.data = &response_context;
+  // Loop over received data
+  while ((num_received = recv(sockfd, buffer, RECEIVE_BUFFER_SIZE, 0)) > 0) {
+    int parsed = http_parser_execute(&parser, &settings, buffer, num_received);
+    // A parsing error occured, or an error in a callback
+    if (parsed < num_received) {
+      return -1;
+    } else if (response_context.status == IOTA_REQUEST_STATUS_ERROR) {
+      return -1;
+    } else if (response_context.status == IOTA_REQUEST_STATUS_DONE) {
+      return RC_OK;
+    }
+    memset(buffer, 0, sizeof(buffer));
+  }
+  if (num_received <= 0) {
+    return -1;
+  }
+  return RC_OK;
+}
+
+static retcode_t send_data_to_iota_service(int sockfd, char const* const data,
+                                           size_t data_length) {
+  char* ptr = (char*)data;
+  while (data_length > 0) {
+    ssize_t num_sent = send(sockfd, ptr, data_length, 0);
+    if (num_sent < 0) {
+      return -1;
+    }
+    ptr += num_sent;
+    data_length -= num_sent;
+  }
+  return RC_OK;
+}
+
+static retcode_t send_headers_to_iota_service(int sockfd,
+                                              http_info_t const* http_settings,
+                                              size_t data_length,
+                                              char const* const path) {
+  char header[100];
+  sprintf(header, "POST %s HTTP/1.1\r\n", path);
+  if (send_data_to_iota_service(sockfd, header, strlen(header)) < 0) {
+    return -1;
+  }
+  sprintf(header, "Host: %s\r\n", http_settings->host);
+  if (send_data_to_iota_service(sockfd, header, strlen(header)) < 0) {
+    return -1;
+  }
+  sprintf(header, "X-IOTA-API-Version: %d\r\n", http_settings->api_version);
+  if (send_data_to_iota_service(sockfd, header, strlen(header)) < 0) {
+    return -1;
+  }
+  sprintf(header, "Content-Type: application/x-www-form-urlencoded\r\n");
+  if (send_data_to_iota_service(sockfd, header, strlen(header)) < 0) {
+    return -1;
+  }
+  sprintf(header, "Content-Length: %lu\r\n", data_length);
+  if (send_data_to_iota_service(sockfd, header, strlen(header)) < 0) {
+    return -1;
+  }
+  if (send_data_to_iota_service(sockfd, "\r\n", 2) < 0) {
+    return -1;
+  }
+  return RC_OK;
+}
+
+retcode_t _iota_service_query(const void* const service_opaque,
```

Comment:
this should alos be static?

Reply:
It's only used for testing, because of the `path` parameter which is not needed for our calls (it defaults to "/"). If I make this function static, it won't be available from the test code, which uses a default value for `path`. Not ideal, but close to a "module private function"